packages/shared/src/retry.ts

```diff
@@ -0,0 +1,108 @@
+type Milliseconds = number;
+
+type RetryOptions = Partial<{
+  /**
+   * The initial delay before the first retry.
+   * @default 125
+   */
+  firstDelay: Milliseconds;
+  /**
+   * The maximum delay between retries.
+   * If set to 0, the delay will increase indefinitely.
+   * @default 0
+   */
+  maxDelay: Milliseconds;
+  /**
+   * The multiplier for the exponential backoff.
+   * @default 2
+   */
+  factor: number;
+  /**
+   * A function to determine if the operation should be retried.
+   * The callback accepts the error that was thrown and the number of iterations.
+   * The iterationsCount variable references the number of retries AFTER attempt
+   * that caused the error and starts at 1 (as in, this is the 1st, 2nd, nth retry).
+   * @default (error, iterationsCount) => iterationsCount < 5
+   */
+  shouldRetry: (error: unknown, iterationsCount: number) => boolean;
+  /**
+   * Controls whether the helper should retry the operation immediately once before applying exponential backoff.
+   * @default true
+   */
+  retryImmediatelyOnce: boolean;
+  /**
+   * The delay for the immediate retry.
+   * @default 100
+   */
+  retryImmediatelyDelay: Milliseconds;
```

Comment:
These two options feel _weird_, I can't put a finger on it but this seems like something that should be only one option and be renamed.

- Retrying immediately already implies that it can only happen once, right? Afterwards it's not immediate anymore
- Technically "immediate" would be a delay of 0 so the current default of 100 is already quite arbitrary. Why should we need to change this from a case by case basis?

So I'd just put those two options into one called `retryImmediate: boolean`

Reply:
Tagging @panteliselef as he was also interested about this - Im also open to suggestions to improve this further. 
There are at least two cases where we might want to immediately retry, before applying the backoff strategy: 
- mission-critical operations that need to be completed as fast as possible 
- transient errors on the network level (think about mobile devices)

This property decouples the immediate retry with the rest of the backoff logic, as its very hard to tune the options using a combination of factor + initialDelay otherwise.

Reply:
I'm going to merge these two into a single `retryImmediate: boolean` with a default delay of `100ms`. If we ever need to change this, we can add a new prop then

Reply:
Done